### ISSUE TITLE
[models] Display problem of vertex colors in the Raylib library for M3D models

### ISSUE THREAD

**Issue by Bigfoot71** [raysan5/raylib#3109]
- [x] I tested it on latest raylib version from master branch
 - [ ] I checked there is no similar issue already reported - ([see issue #2855](https://github.com/raysan5/raylib/issues/2855))
 - [x] My code has no errors or misuse of raylib

### Issue description

I'm having a problem using raylib to display M3D models with vertex colors. The vertex colors are not displayed correctly in my scene, while the official [m3dview](https://gitlab.com/bztsrc/model3d) viewer manages to display the vertex colors correctly.

### Environment

Desktop, Windows 10, OpenGL 3.3:
```
INFO: GL: OpenGL device information:
INFO:     > Vendor:   NVIDIA Corporation
INFO:     > Renderer: NVIDIA GeForce GTX 980/PCIe/SSE2
INFO:     > Version:  3.3.0 NVIDIA 535.98
INFO:     > GLSL:     3.30 NVIDIA via Cg compiler
```

### Issue Screenshot

![m3dview-1 m3d](https://github.com/raysan5/raylib/assets/90587919/3fbd25f6-406e-46a7-bbc8-e2c9e9218692)
![raylib-1 m3d](https://github.com/raysan5/raylib/assets/90587919/e6a128b3-5784-48a1-940c-64a02baca0bb)

### Code Example

[test-m3d.zip](https://github.com/raysan5/raylib/files/11734283/test-m3d.zip)

**Comment by Bigfoot71** [comment]
@chriscamacho Effectively, the vertex colors are just not loading, so I will have to write a solution myself, thanks for your answer.

**Comment by Bigfoot71** [comment]
I am back to the problem I presented here, and it turns out that you simply needed to add the following to `LoadM3D` in `rmodels.c` after something like line [5625](https://github.com/raysan5/raylib/blob/master/src/rmodels.c#L5625):

```
model.meshes[k].colors = (unsigned char *)RL_CALLOC(model.meshes[k].vertexCount*4, sizeof(unsigned char));
```

And the rest of the code was already implemented here: [L2655](https://github.com/raysan5/raylib/blob/master/src/rmodels.c#L5655)

And the model displays its beautiful vertex colors! :D

**Comment by raysan5** [comment]
@Bigfoot71 Feel free to send a PR

**Comment by raysan5** [comment]
@bztsrc please, could you take a look to this issue?

**Comment by bztsrc** [comment]
> please, could you take a look to this issue?

Sure!

> it seems that there is no direct, simple, and reliable way to know if there are vertex colors in the loaded m3d model

No, because M3D can store both vertex colors and materials (with textures). You can have a model with colors, that's only used by loaders that do not load materials; and never be shown by loaders that support materials and textures.

> In the m3d model, color is not an array itself like in raylib; it is an unsigned int

Of course, because color is a 32-bit packed pixel with separate RGBA channels.

> the only solution I can think of would be to allocate mesh.colors by default 

You don't need mesh.colors at all. Even if the model uses a color-map, the SDK automatically converts that into 32-bit RGBA packed pixels, so all you can see is a consistent, true-color compatible code.

And this is intended. You shouldn't know if the compression algorithm optimized the colors into a palette or not; you should get the same RGBA pixels for the vertex colors no matter how those were stored in the file.

> I am back to the problem I presented here, and it turns out that you simply needed to add the following to LoadM3D in rmodels.c after something like line 5625:

Now this is a different thing. This relates to the VertexBuffer that the GPU uses. Yes, you could always allocate the colors array if you want, that won't hurt.

> If we allocate mesh.colors by default but there is no vertex color in the model the function will return 0x00000000 for all vertex colors, which can be problematic.

No. It would fill up mesh.colors with black pixels in worst case, this isn't problematic at all. BTW, even if the allocation fails and mesh.colors indeed becomes 0x0000000, there are checks to avoid NULL dereference, so that wouldn't be a problem either.

To sum it up, I believe adding https://github.com/raysan5/raylib/issues/3109#issuecomment-1634843322 would waste some memory for vertex color-less models, but otherwise I see no harm at all. I think it should be merged.

Cheers,
bzt

**Comment by raysan5** [comment]
> To sum it up, I believe adding https://github.com/raysan5/raylib/issues/3109#issuecomment-1634843322 would waste some memory for vertex color-less models, but otherwise I see no harm at all. I think it should be merged.

I think this is a key point, most models nowaday do not use vertex colors, allocating vertex colors for every loaded model would be, in most cases, a waste of memory.

**Comment by bztsrc** [comment]
> I think this is a key point, most models nowaday do not use vertex colors, allocating vertex colors for every loaded model would be, in most cases, a waste of memory.

Yes, you're right, it's not common. Here's a proper solution: [rmodels.c.zip](https://github.com/raysan5/raylib/files/12310650/rmodels.c.zip)

In this, when the number of vertices counted for a material, I've added an extra check to see if all vertex colors are fully transparent black or not. The mesh.colors array is only allocated if no materials is defined or there's at least one non fully transparent black color. I've tested this, it works perfectly with normal models (material only), models without materials (which supposed to use colors), and models with both materials and colors (in this case raylib's shader seems to mix the vertex color and the texture's color).
![raylibm3d](https://github.com/raysan5/raylib/assets/12779069/3567fe3f-3559-4189-a574-d68c97ee50c6)

NOTE: I've also added the materialCount fix from https://github.com/raysan5/raylib/issues/3177 to this version.

Cheers,
bzt

ps: I'm terribly sorry for creating an attachment instead of a PR, but I have troubles with github ever since it changed to this JS heavy interface. Sorry about this!

**Comment by Bigfoot71** [comment]
I believe there was a misunderstanding with my previous message, I must have expressed myself poorly. It's not a big deal though.

So, we are indeed required to reiterate through the faces initially to determine if there are vertex colors?

There is no other more direct way? Perhaps it's better to keep it as it was before, right?
This could be heavy for the rare cases where we would need it?

**Allowing the choice to allocate mesh.colors or not was therefore appropriate in a certain sense?**

**Comment by raysan5** [comment]
@bztsrc thank you very much for the review! just merged your changes!

 > I'm terribly sorry for creating an attachment instead of a PR, but I have troubles with github ever since it changed to this JS heavy interface. Sorry about this!

No problem! I agree this new interface is really slow!
